### PR TITLE
Saas 17 valuation merge

### PR DESCRIPTION
**Pull request by jco-odoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by sle-odoo** [comment]
@hbto

**Comment by hbto** [comment]
thanks for the ping @sle-odoo

**Comment by nhomar** [comment]
it should be lovely if this land in saas-17 :+1:

**Comment by nhomar** [comment]
RED ALERT:

average_price = fields.Float(
        'Average Cost', compute='_compute_average_price',
        digits=dp.get_precision('Product Price'), groups="base.group_user",
        help="Average cost of the product, in the default unit of measure of the product.")

this field MUST by multicompany aware, if not this feature is useless (as v5, v6, v7, v8, v9.....).

**Comment by gdgellatly** [comment]
Can you have a computed multi company field? That would be a functional
functional field in old money.

File: addons/stock_landed_costs/models/stock_landed_cost.py [hunk 1]
```diff
@@ -98,13 +93,16 @@ def button_validate(self):
...
+                old_value = line.move_id.product_id.stock_value
+                line.move_id.write({'landed_cost_value': line.move_id.landed_cost_value + cost_to_add,
+                                    'value': line.move_id.value + cost_to_add, 
+                                    'price_unit': (line.move_id.value + cost_to_add) / line.move_id.product_qty,})
+                line.move_id.replay()
+                line.move_id.product_id._compute_stock_value()
+                new_value = line.move_id.product_id.stock_value
+                diff = new_value - old_value
+                extra_cost = diff - cost_to_add
+                line._create_accounting_entries(move, diff, extra_cost)
```
Comment (hbto) [inline addons/stock_landed_costs/models/stock_landed_cost.py]: Landed cost is only to be applied on incoming moves,

No validation is performed when computing or when validating the Landed Costs.

https://github.com/Vauxoo/addons-vauxoo/blob/8.0/stock_landed_costs_average/model/stock_landed_costs.py#L432-L434

File: addons/stock_account/models/stock.py [hunk 2]
```diff
@@ -75,8 +76,8 @@ class StockMove(models.Model):
     last_done_remaining_qty = fields.Float()
     last_done_qty = fields.Float()
```
Comment (hbto) [inline addons/stock_account/models/stock.py]: Will there ever be a relation between stock_move and account_move_line?

https://github.com/Vauxoo/addons-vauxoo/blob/8.0/stock_move_entries/model/stock_move_entries.py#L51-L52

https://github.com/Vauxoo/addons-vauxoo/blob/8.0/stock_move_entries/model/stock_move_entries.py#L36-L38

if we want to keep a nitty-gritty detail of accounting valuation this is a must-have.
I know that currently we have a post in the mail_thread in the stock_move/stock_picking
but it is not fully helpful when try to find the right accounting valuation of a product.

File: addons/stock_landed_costs/models/stock_landed_cost.py [hunk 3]
```diff
@@ -323,47 +321,24 @@ def _create_account_move_line(self, move, credit_account_id, debit_account_id, q
             credit_line['debit'] = -diff
         AccountMoveLine.create(debit_line)
         AccountMoveLine.create(credit_line)
-
-        # Create account move lines for quants already out of stock
-        if qty_out > 0:
+        
+        # In anglo-saxon we need to deduct the COGS
+        if self.env.user.company_id.anglo_saxon_accounting and float_compare(extra_cost, 0.0, precision_rounding=self.env.user.company_id.currency_id.rounding) != 0.0:
```
Comment (hbto) [inline addons/stock_landed_costs/models/stock_landed_cost.py]: Excellent thanks for incorporating this.

**Comment by sle-odoo** [comment]
@nhomar yes, this is supposed to land in saas-17

File: addons/stock_landed_costs/models/stock_landed_cost.py [hunk 4]
```diff
@@ -323,47 +321,24 @@ def _create_account_move_line(self, move, credit_account_id, debit_account_id, q
             credit_line['debit'] = -diff
         AccountMoveLine.create(debit_line)
```
Comment (hbto) [inline addons/stock_landed_costs/models/stock_landed_cost.py]: This could be a lengthy process:

aml_model.with_context(check_move_validity=False)

File: addons/stock_landed_costs/models/stock_landed_cost.py [hunk 5]
```diff
@@ -132,7 +130,7 @@ def get_valuation_lines(self):
 
         for move in self.mapped('picking_ids').mapped('move_lines'):
             # it doesn't make sense to make a landed cost for a product that isn't set as being valuated in real time at real cost
-            if move.product_id.valuation != 'real_time' or move.product_id.cost_method not in ('average', 'fifo'):
+            if move.product_id.cost_method not in ('average', 'fifo'):
```
Comment (hbto) [inline addons/stock_landed_costs/models/stock_landed_cost.py]: When the valuation is the last record, assign line.price_unit less the accumulated value in value_split

https://github.com/Vauxoo/addons-vauxoo/blob/8.0/stock_landed_costs_average/model/stock_landed_costs.py#L592-L602

File: addons/stock_account/models/product.py [hunk 6]
```diff
@@ -114,11 +109,6 @@ class ProductProduct(models.Model):
...
         'Value', compute='_compute_stock_value')
 
-    @api.onchange('type')
-    def onchange_type_valuation(self):
-        # TO REMOVE IN MASTER
-        pass
-
     @api.multi
     def do_change_standard_price(self, new_price, account_id):
         """ Changes the Standard Price of Product and creates an account move accordingly."""
```
Comment (hbto) [inline addons/stock_account/models/product.py]: There are location for which this product could not have any qty_available,
`for location in locations:`

It would be better if this is done by first checking which locations actually
have this product.

https://github.com/Vauxoo/addons-vauxoo/blob/8.0/product_extended_variants/model/template.py#L101-L113

@moylop260 @nhomar Do you remember this performance problem?

File: addons/stock_account/models/product.py [hunk 7]
```diff
@@ -200,7 +195,8 @@ def _get_candidates_move(self):
             ('location_dest_id.usage', 'in', ('transit', 'internal')),
             ('location_id.usage', 'not in', ('transit', 'internal')),
             ('remaining_qty', '>', 0),
-            ('state', '=', 'done')
+            ('state', '=', 'done'), 
+            ('company_id', '=', self.company_id.id)
         ], order='date, id') #TODO: case where 
         return candidates
```
Comment (hbto) [inline addons/stock_account/models/product.py]: on `_compute_average_price` is useful to have a meaningful value of `average_price` != 0 
because if for some reason the product get depleted and you want to adjust inventory, and 
it turns out that there was products available, you then end up with adjusting a product
with valuation zero, 

https://github.com/Vauxoo/addons-vauxoo/blob/8.0/stock_card/model/stock_card.py#L460-L463

We have witnessed this scenario with our customers.

File: addons/stock_account/models/product.py [hunk 8]
```diff
@@ -73,11 +73,6 @@ def _is_cost_method_standard(self):
     def _set_cost_method(self):
         return self.write({'property_cost_method': self.cost_method})
```
Comment (hbto) [inline addons/stock_account/models/product.py]: Has any check been made when a product changes from one category to another with different cost_method?

https://github.com/odoo/odoo/pull/18529/files#diff-e64c2b05f2fa7cb4b77eada8a18bea6bR258

Just the way it has been made at category level.

File: addons/stock_account/models/product.py (hunk 6 above)
Comment (jco-odoo) [inline addons/stock_account/models/product.py]: Valuation is now changed on the category.  Indeed, adding the product_id on the account move would not be bad.  (if not already, saas-18)

File: addons/stock_account/models/product.py [hunk 9]
```diff
@@ -224,35 +220,80 @@ def _compute_stock_value(self):
...
+        qty_todo = self.with_context(internal=True).qty_available
+        for candidate in candidates:
+            if qty_todo > candidate.product_qty:
+                candidate.remaining_qty = candidate.product_qty
+                qty_todo -= candidate.product_qty
+                candidate.price_unit = price
+                candidate.value = price * candidate.product_qty
+            else:
+                candidate.remaining_qty = qty_todo
+                candidate.value = candidate.value * (candidate.product_qty - qty_todo) / candidate.product_qty + price * qty_todo
```
Comment (jco-odoo) [inline addons/stock_account/models/product.py]: Indeed.

File: addons/stock_account/models/stock.py (hunk 2 above)
Comment (jco-odoo) [inline addons/stock_account/models/stock.py]: An idea for saas-18.

File: addons/stock_account/models/stock.py [hunk 10]
```diff
@@ -87,83 +88,211 @@ def _get_price_unit(self):
...
+        return [('product_id', '=', self.product_id.id), 
+                 ('state', '=', 'done'), 
+                 ('company_id', '=', self.company_id.id),
+                 '|', '&', ('location_id.usage', 'in', ('internal', 'transit')), 
+                    ('location_dest_id.usage', 'not in', ('internal', 'transit')), 
+                    '&', ('location_id.usage', 'not in', ('internal', 'transit')), 
+                    ('location_dest_id.usage', 'in', ('internal', 'transit'))]
+        
+    def _is_in_move(self):
+        return self.location_id.usage not in ('internal', 'transit') and self.location_dest_id.usage in ('internal', 'transit')
```
Comment (jco-odoo) [inline addons/stock_account/models/stock.py]: If it is within the company, it should not be considered in/out.  I was thinking about changing it in: not self.location_id.company_id and self.location_dest_id.company_id == current_company.  Otherwise it is not company_id and location transit (the condition becomes more complicated)

Comment (hbto) [inline addons/stock_account/models/stock.py]: I know what you mean by `If it is within the company` because I and my colleagues hit this same issue a long time ago, the thing with merchandise going into transit, is that it creates Journal Entries when going out and coming in, if that is the case we have to be consistent with the value of the merchandise (if creating JE), or we have to be consider that merchandise going through inter-transit/same-company does not create JEs, and as you said, it is complex issue.

File: addons/stock_account/models/product.py (hunk 7 above)
Comment (jco-odoo) [inline addons/stock_account/models/product.py]: Could be the price after the last in then?  (or the price used for the last out)

File: addons/stock_account/models/product.py (hunk 8 above)
Comment (jco-odoo) [inline addons/stock_account/models/product.py]: Let us me think that if we change on the category, we should not change the products which have property_cost_method defined.  And we could apply the same logic here by product.  (although the field is not shown in the views)

Comment (nhomar) [inline addons/stock_account/models/product.py]: Yes

File: addons/stock_account/models/product.py (hunk 6 above)
Comment (nhomar) [inline addons/stock_account/models/product.py]: @jco-odoo If you ask me an opinion, I personally prefer stay with this unmerged until saas-18 with full review test than left known technical debts that will clearly give problems in saas-17.

 prefer have this very first.

But if this become a perfonrmance problem this can be fixed in the same stable (-17) after merge, because is not a matter of logic.

The performance problem WILL be there.

If you have for example a little business with 20 locations (that's little size if you want to have precise control of locations) you can see the problem easy.

And about the product in the account_move_line AFAICR historically this was always there even on the name, but we must include the traceability data (even if reports to use the daa come afterwards).

File: addons/stock_account/models/product.py [hunk 11]
```diff
@@ -224,35 +220,80 @@ def _compute_stock_value(self):
...
+            ('location_id.usage', 'not in', ('transit', 'internal')),
+            ('state', '=', 'done'), 
+            ('company_id', '=', self.company_id.id)], order='date desc, id desc')
+        qty_todo = self.with_context(internal=True).qty_available
+        for candidate in candidates:
+            if qty_todo > candidate.product_qty:
+                candidate.remaining_qty = candidate.product_qty
+                qty_todo -= candidate.product_qty
+                candidate.price_unit = price
+                candidate.value = price * candidate.product_qty
```
Comment (nhomar) [inline addons/stock_account/models/product.py]: @jco-odoo this is important, and IMHO this can be done now before merge.

File: addons/stock_account/models/product.py (hunk 7 above)
Comment (hbto) [inline addons/stock_account/models/product.py]: @jco-odoo Yes, the last out.

File: addons/stock_account/models/product.py (hunk 6 above)
Comment (jco-odoo) [inline addons/stock_account/models/product.py]: It would be simpler to simply remove or "group by" the locations.  Do you have knowledge of a lot of people putting different accounts on locations?

Comment (hbto) [inline addons/stock_account/models/product.py]: location.valuation_in_account_id and location.valuation_out_account_id are only used when products are moved to/from production or inventory locations , isn't it. When adjusting valuation of products by using this method, accounts are coming from template/category accounts.

So for locations with type `internal` (or `transit` following your same approach of thinking transit is like internal) no accounts from there are used. 
At this point only locations with those two types are to be checked for qty_available,

Thus grouping is ok and it won't affect those locations using accounts.

File: addons/purchase/test/fifo_price.yml [hunk 12]
```diff
@@ -343,7 +343,7 @@
     self.env['stock.immediate.transfer'].create({'pick_id': picking.id}).process()
     original_out_move = self.env['stock.picking'].browse(ref('outgoing_fifo_shipment_neg')).move_lines[0]
     assert original_out_move.remaining_qty == 50.0, 'On the out move of 100, 50 still needs to be matched'
-    assert original_out_move.value == 2500, 'Value of the move should be 2500'
+    assert original_out_move.value == -2500, 'Value of the move should be 2500'
```
Comment (sle-odoo) [inline addons/purchase/test/fifo_price.yml]: note: this change is because the negative stock was is wrongly implemented in saas-17 at the moment (in all cases, an out move should be negative)

File: addons/sale_stock/models/account_invoice.py [hunk 13]
```diff
@@ -52,7 +52,7 @@ def _compute_average_price(self, qty_done, quantity, moves):
                 qty_to_consider = invoiced_qty - qty_done
             qty_to_consider = min(qty_to_consider, quantity - qty_delivered)
             qty_delivered += qty_to_consider
-            average_price_unit = (average_price_unit * (qty_delivered - qty_to_consider) + move.price_unit * qty_to_consider) / qty_delivered
+            average_price_unit = (average_price_unit * (qty_delivered - qty_to_consider) - (move.value / move.product_qty) * qty_to_consider) / qty_delivered
```
Comment (sle-odoo) [inline addons/sale_stock/models/account_invoice.py]: to clarify:
- move.price_unit
- move._get_price_unit()
- move.value

Actually, move._get_price_unit is intended to set the value on IN move

action_done on out move will write move.price_unit according to the cost method

so this change is probably useless

File: addons/stock/models/product.py [hunk 14]
```diff
@@ -76,7 +76,7 @@ class Product(models.Model):
     reordering_min_qty = fields.Float(compute='_compute_nbr_reordering_rules')
     reordering_max_qty = fields.Float(compute='_compute_nbr_reordering_rules')
 
-    @api.depends('stock_quant_ids', 'stock_move_ids')
+    @api.depends('stock_move_ids.product_qty', 'stock_move_ids.state')
     def _compute_quantities(self):
```
Comment (sle-odoo) [inline addons/stock/models/product.py]: note: why this change: no need to check on both quants and moves, and stock.move will be used later in the method to compute the theoretical quantities, so just keep the stock move dependency

File: addons/stock/models/product.py [hunk 15]
```diff
@@ -154,9 +154,14 @@ def _get_domain_locations(self):
...
         Possible parameters are shop, warehouse, location, force_company, compute_child
         '''
-        # TDE FIXME: clean that brol, context seems overused
         Warehouse = self.env['stock.warehouse']
 
+        if self.env.context.get('internal', False):
+            company_id = self.env.user.company_id.id
+            return ([('location_id.usage', 'in', ('internal', 'transit')), ('company_id', '=', company_id)], 
+                    [('location_id.usage', 'not in', ('internal', 'transit')), ('location_dest_id.usage', 'in', ('internal', 'transit')), ('company_id', '=', company_id)],
+                    [('location_id.usage', 'in', ('internal', 'transit')), ('location_dest_id.usage', 'not in', ('internal', 'transit')), ('company_id', '=', company_id)])
```
Comment (sle-odoo) [inline addons/stock/models/product.py]: could be simplified in location_id.company_id company_id (probably)

File: addons/stock_account/models/stock.py [hunk 16]
```diff
@@ -150,20 +177,29 @@ def replay(self):
...
+            except for standard cost price, where it will put the standard price.  
+            For FIFO, it will also check if there was negative stock and compensate the corresponding out moves.   
+            For outgoing moves, average price, will apply the average price formula, FIFO will search the 
+            corresponding incoming moves where it can reduce the remaining_qty while standard takes the cost from the product
+        """
         qty_available = {}
         for move in self:
-            #Should write move.price_unit here maybe, certainly on incoming
+            #Should write move.price_unit here maybe, certainly on incoming moves
             if move.product_id.cost_method == 'average':
```
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: why this change: average price should look for every product available in the company, even if they're in transit between two warehouses

File: addons/stock_account/models/product.py [hunk 17]
```diff
@@ -172,6 +172,11 @@ def _get_latest_cumulated_value(self, not_move=False):
         domain = [
             ('product_id', '=', self.id),
             ('state', '=', 'done'),
+            ('company_id', '=', self.company_id.id),
```
Comment (sle-odoo) [inline addons/stock_account/models/product.py]: note: rename no_move in exclude

File: addons/stock_account/models/product.py [hunk 18]
```diff
@@ -172,6 +172,11 @@ def _get_latest_cumulated_value(self, not_move=False):
         domain = [
             ('product_id', '=', self.id),
             ('state', '=', 'done'),
+            ('company_id', '=', self.company_id.id), 
+            '|', '&', ('location_id.usage', 'in', ('internal', 'transit')), 
+            ('location_dest_id.usage', 'not in', ('internal', 'transit')), 
+            '&', ('location_id.usage', 'not in', ('internal', 'transit')), 
+            ('location_dest_id.usage', 'in', ('internal', 'transit'))
```
Comment (sle-odoo) [inline addons/stock_account/models/product.py]: what about inter company transit?
so we should look for location_dest_id.company (inter company transit have no company set on their transit location)

File: addons/stock_account/models/stock.py [hunk 19]
```diff
@@ -87,27 +87,54 @@ def _get_price_unit(self):
             move = self.search([('product_id', '=', self.product_id.id),
                          ('state', '=', 'done'), 
                          ('location_id.usage', '=', 'internal'), 
-                         ('location_dest_id.usage', '!=', 'internal')], order='date desc', limit=1)
+                         ('location_dest_id.usage', '!=', 'internal'),
+                         ('company_id', '=', self.company_id.id)], order='date desc', limit=1)
```
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: should probably expose this domain, as it's the same domain as _get_candidates_out_move but with a different order and limit

File: addons/stock_account/models/stock.py [hunk 20]
```diff
@@ -87,27 +87,54 @@ def _get_price_unit(self):
...
-                     ('date', '>',  self.date), 
-                     ('product_id', '=', self.product_id.id)])
+                     ('date', '>=',  self.date),
+                     ('id', '>', self.id), 
+                     ('product_id', '=', self.product_id.id), 
+                     ('company_id', '=', self.company_id.id),
+                     '|', '&', ('location_id.usage', 'in', ('internal', 'transit')), 
+                    ('location_dest_id.usage', 'not in', ('internal', 'transit')), 
+                    '&', ('location_id.usage', 'not in', ('internal', 'transit')), 
+                    ('location_dest_id.usage', 'in', ('internal', 'transit'))])
```
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: should call get_all_domain

File: addons/stock_account/models/stock.py [hunk 21]
```diff
@@ -87,27 +87,54 @@ def _get_price_unit(self):
...
         return self.product_id.standard_price
 
     def _update_future_cumulated_value(self, value):
         self.ensure_one()
         moves = self.search([('state', '=', 'done'), 
-                     ('date', '>',  self.date), 
-                     ('product_id', '=', self.product_id.id)])
+                     ('date', '>=',  self.date),
+                     ('id', '>', self.id), 
+                     ('product_id', '=', self.product_id.id),
```
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: always compare date, compare id only if date are equal

File: addons/stock_account/models/stock.py [hunk 22]
```diff
@@ -87,27 +87,54 @@ def _get_price_unit(self):
...
                          ('location_id.usage', '=', 'internal'), 
-                         ('location_dest_id.usage', '!=', 'internal')], order='date desc', limit=1)
+                         ('location_dest_id.usage', '!=', 'internal'),
+                         ('company_id', '=', self.company_id.id)], order='date desc', limit=1)
             if move:
                 return move.price_unit or self.product_id.standard_price
         return self.product_id.standard_price
 
     def _update_future_cumulated_value(self, value):
         self.ensure_one()
```
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: negative stock in fifo: (out something you don't have in stock) do not assign a value (0), they will be compensated by this method when a receipt compensates

File: addons/stock_account/models/stock.py (hunk 10 above)
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: @hbto i think it's the case in v8-10, if you're in transit between warehouses of the same compnay, journal entries won't be created. We have to that here (by using location.compaby_id on the above domain, and not juste location usage)

File: addons/stock_account/models/stock.py [hunk 23]
```diff
@@ -150,20 +177,29 @@ def replay(self):
...
                     if not move.price_unit:
                         move.price_unit = move._get_price_unit()
-                    move.value = move.price_unit * move.product_qty
-                    move.cumulated_value = move.product_id._get_latest_cumulated_value(not_move=move) + move.value
-                    move.remaining_qty = move.product_qty
+                    move_value = move.price_unit * move.product_qty
+                    move.write({'value': move_value, 
+                                'cumulated_value': move.product_id._get_latest_cumulated_value(not_move=move) + move_value, 
+                                'remaining_qty': move.product_qty, 
+                                'last_done_qty': move.product_id.with_context(internal=True).qty_available,})
```
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: note: used when you edit move in the past and the cost method is average
not used for fifo?

Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: acutally last_done_qty is not used in this commit, it'll be used by a further one that fix average edit in the past

File: addons/stock_account/models/stock.py [hunk 24]
```diff
@@ -173,18 +209,22 @@ def action_done(self):
                                 qty_taken_on_candidate = candidate.remaining_qty
                             else:
                                 qty_taken_on_candidate = qty_to_take
-                            candidate.remaining_qty -= qty_taken_on_candidate
                             move.remaining_qty -= qty_taken_on_candidate
                             qty_to_take -= qty_taken_on_candidate
-                            candidate.value += move.price_unit * qty_taken_on_candidate
-                            candidate.cumulated_value += move.price_unit * qty_taken_on_candidate
+                            candidate_value = candidate.value - move.price_unit * qty_taken_on_candidate
```
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: this line is actually the fix the fifo with negative stock

File: addons/stock_account/models/stock.py [hunk 25]
```diff
@@ -173,18 +209,22 @@ def action_done(self):
...
+                            candidate.write({'value': candidate_value, 
+                                             'cumulated_value': candidate.cumulated_value - move.price_unit * qty_taken_on_candidate, 
+                                             'price_unit': - (candidate_value / candidate.product_qty),
+                                             'last_done_move_id': move.id,
```
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: for an out, it's the last in the out compensated

File: addons/stock_account/models/stock.py [hunk 26]
```diff
@@ -201,23 +241,26 @@ def action_done(self):
...
-                    if last_candidate:
-                        move.last_done_move_id = last_candidate.id
-                        move.last_done_remaining_qty = last_candidate.remaining_qty
-                    if qty_to_take > 0:
-                        move.remaining_qty = qty_to_take # In case there are no candidates to match, put standard price on it
-                    move.value = -tmp_value
-                    move.cumulated_value = move.product_id._get_latest_cumulated_value(not_move=move) + move.value
-                    move.last_done_qty = move.product_id.qty_available
+                    move.write({'value': -tmp_value,
+                                'price_unit': tmp_value / move.product_qty,
```
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: before, price_unit was not set on out move

File: addons/stock_landed_costs/models/stock_landed_cost.py [hunk 27]
```diff
@@ -323,47 +321,24 @@ def _create_account_move_line(self, move, credit_account_id, debit_account_id, q
...
+                              account_id=credit_account_id)
             credit_line = dict(base_line,
-                               name=(self.name + ": " + str(qty_out) + _(' already out')),
-                               quantity=qty_out,
-                               account_id=debit_account_id)
-            diff = diff * qty_out / self.quantity
-            if diff > 0:
-                debit_line['debit'] = diff
-                credit_line['credit'] = diff
+                               name=(self.name + _(' already out')),
```
Comment (xmo-odoo) [inline addons/stock_landed_costs/models/stock_landed_cost.py]: Should use string formatting so translations can do the entire thing at once, in case e.g. they need to change the ordering of the phrase.

**Comment by hbto** [comment]
There is an issue with the process of creating Journal Entries at:
- Purchase reception and Purchase Billing,
- Sales delivery and Sales Invoicing,

At least when anglo_saxon is enable

This issue has become kind of cumbersome to some of our clients and the community in general which try to keep everything up-to-date and not just slashing adjustments into the accounting journals to try to keep the accounting entries reconciled.

We have been trying for some time to mitigate the lack of a tools native in odoo, with a customized module, https://github.com/Vauxoo/addons-vauxoo/blob/8.0/stock_accrual_report

where we mix account_anglo_saxon_stock_move_purchase, account_anglo_saxon_stock_move_sale 
with the goal to try to keep the track and reconcile the journal entry lines that were created using stock_in and stock_out accounts.

There is a lengthy conversation at the mailing list regarding this issue.
https://odoo-community.org/groups/accounting-28/accounting-31466?mode=thread

I can see that here you are trying to fix the issue of the changing price when invoicing to customer, because the COGS can change in between delivery and invoicing.
https://github.com/odoo-dev/odoo/blob/2bdd141d63639aa86e06e3d9eff8a443fe993a2c/addons/sale_stock/models/account_invoice.py#L40

Every solution has its advantages and disadvantages but the issue remains the same 
as there could be many layers of inheritance we can accomplish the goals but at the cost of performance where there are many purchase/sale lines and invoice lines involve to reconcile.

If you have foresee any approach we at the community would be glad to see what have you figure out, if not please open a channel where we can discuss the issue.

Best Regards

File: addons/stock_account/models/product.py [hunk 28]
```diff
@@ -39,18 +39,23 @@ class ProductTemplate(models.Model):
...
     average_price = fields.Float(
-        'Average Cost', compute='_compute_average_price',
+        'Average Cost', compute='_compute_fifo_average_price',
         digits=dp.get_precision('Product Price'), groups="base.group_user",
-        help="Average cost of the product, in the default unit of measure of the product.")
+        help="This is the average cost of this product and can be used to value the stock of this product. ")
+    fifo_price = fields.Float(
+        'FIFO Cost', compute='_compute_fifo_average_price',
+        digits=dp.get_precision('Product Price'), groups="base.group_user",
+        help="This is the cost based on the last products leaving the stock.")
```
Comment (nhomar) [inline addons/stock_account/models/product.py]: This means that if you just received the product and did not deipatch any sale you can not have a "Fifo BAsed cost"?.

At least what I understand from the help text.

I want to point that the very first cost of the product is generally the one used to set the sale price.

Comment (jco-odoo) [inline addons/stock_account/models/product.py]: Maybe we need to call it FIFO return price.  
For the very first cost we would use the standard cost then.  (and not adapt the standard cost anymore as in v10)

Comment (nhomar) [inline addons/stock_account/models/product.py]: If this is what means: "Maybe we need to call it FIFO return price." it is worst my friend,

The FIFO return price should follow this logic:

fifo on the origin (if origin provided) or "oldest fifo delivered" be carefull with that.

And remember always that the retunr must be in the cost that goes out if provided.

Regards.

File: addons/stock_account/models/stock.py [hunk 29]
```diff
@@ -75,95 +76,218 @@ class StockMove(models.Model):
...
+            move.write({'cumulated_value': last_cumulated_value, 
+                        'last_done_qty': last_done_qty_available})
+
+    def replay_fifo(self):
+        """
+            We try to find an outgoing move before this one that did not 
+            have a negative stock when it was done and has a corresponding in move. 
+            By taking the later in moves and out moves, we recalculate the remaining qties and values on the moves
+            Afterwards, we take both in and out moves chronologically and recalculate cumulated value and qty done
+        """
```
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: ensure one

File: addons/stock_account/models/stock.py [hunk 30]
```diff
@@ -75,95 +76,218 @@ class StockMove(models.Model):
...
+        out_domain = self._get_out_domain()
+        out_domain += ['|', ('date', '<', self.date),
+                     '&', ('date', '=', self.date), ('id', '<', self.id), 
+                     ('last_done_move_id', '!=', False),
+                     ('last_done_qty', '>=', 0.0)]
+        start_move = self.search(out_domain, limit=1, order='date desc, id desc')
+        # normally, this will be the case as last_done_qty > 0.0
+        use_start_move = start_move and start_move.last_done_move_id and (start_move.last_done_move_id.date < start_move.date or (start_move.last_done_move_id.date == start_move.date and start_move.last_done_move_id.id < start_move.id))
```
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: start_move.last_done_move_id.date < start_move.date excludes negative moves that were compensated

File: addons/stock_account/models/stock.py [hunk 31]
```diff
@@ -75,95 +76,218 @@ class StockMove(models.Model):
...
+                        total_qty = 0
+                out_move.write({'last_done_remaining_qty': last_remaining_qty, 
+                                'last_done_move_id': last_in_move.id, 
+                                'value': -total_value,
+                                'price_unit': total_value / out_move.product_qty,})
+                out_move.remaining_qty = total_qty
+            for move in in_moves_needed:
+                move.remaining_qty = move.product_qty
+        # Recalculate last_done_qty and cumulated value on next moves
+        domain = self._get_all_domain()
```
Comment (sle-odoo) [inline addons/stock_account/models/stock.py]: actually _get_in_and_out_domain .. :)

**Comment by hbto** [comment]
Hello @jco-odoo @sle-odoo 

Have you tested stock_account with new average computation for returns from customers:
So far it has failed to pass this tests in
https://github.com/vauxoo-dev/odoo-average/blob/master/avco_tests/tests/test_stock_account_unfuck.py#L149-L158

I have tested master / saas-17-valuation-merge / saas-17-valuation-sle
so far the three branches are failing against some tests
https://travis-ci.org/vauxoo-dev/odoo-average

Could you review tests?, I could have left something loose 

Best Regards

**Comment by sle-odoo** [comment]
Hi @hbto, thanks for keeping an eye on these branches. I still hope to merge stock-valuation-sle by the end of the week but i want to simplify a lot the structure of the code, especially the replay method, maybe make one `_run_fifo` method that we can call on any move, be it a new one in `action_done` or an old one when editing a picking in the past (this adds a lot of complexity by the way), or even when applying a landed cost or validating a vendor bill with a price different than the purchase. I added some [tests](https://github.com/odoo-dev/odoo/blob/ca82ee0fe6b51bf32b2391f5d4e90ea4197fd9b1/addons/stock_account/tests/test_stockvaluation.py) in this branch and made some corrections for the negative stock AVCO.

So, what would be the correct behaviour when returning goods in AVCO? I have some trouble understanding the testcases you linked.

**Comment by hbto** [comment]
Hi @sle-odoo I have made a video so that it could be easier to follow the Tests
https://youtu.be/qc-AvzStSzY

**Comment by hbto** [comment]
# The Overall Expected Result:

In the image below is shown a **Stock Card** for Product `VX Victrola` with
transactions as follows:

1. Purchase Order for 10 items @ $100 Cost Price - Long Blue Arrow Tail.
2. Sale Order for 4 items @ $100 (CoGS).
3. Sale Order for 5 items @ $100 (CoGS) - Black Arrow Tail.
4. Purchase Order for 3 items @ $300 Cost Price - Red Arrow Tail.
5. Sale Order for 3 items @ $250 (CoGS).
6. Customer Returns 1 items @ $100.00 - Cost of Good when it was sold - Black
   Arrow Head - **Average changes**.
7. Purchase Order for 5 items @ $220 Cost Price - Short Blue Arrow Tail.
8. Purchase Order for 3 items @ $400 Cost Price - Orange Arrow Tail.
9. Return to Supplier 5 items from transaction `7` @ $220 - Short Blue Arrow
   Head - **Average changes**.
10. Return to Supplier 1 items from transaction `4` @ $300 - Red Arrow Head -
    **Average changes**.
11. Return to Supplier 1 items from transaction `1` @ $100 - Long Blue Arrow
    Head - **Average changes**.
11. Return to Supplier 3 items from transaction `8` @ `average` - Orange Arrow
    Head - **Average is kept the same**.

Stock Card

Now in the next image we are showing you that Accounting can be hand-in-hand
with logistics.

All the Journal Items are the ones for the Stock Valuation Account for the
product being presented previously.

Journal Items

Taken from https://github.com/Vauxoo/addons-vauxoo/blob/8.0/stock_account_unfuck/README.md

File: addons/stock_account/models/product.py [hunk 32]
```diff
@@ -220,39 +221,83 @@ def _compute_stock_value(self):
...
         company_dependent=True, copy=True, required=True,
-        help="If perpetual valuation is enabled for a product, the system "
-             "will automatically create journal entries corresponding to "
-             "stock moves, with product price as specified by the 'Costing "
-             "Method'. The inventory variation account set on the product "
-             "category will represent the current inventory value, and the "
-             "stock input and stock output account will hold the counterpart "
-             "moves for incoming and outgoing products.")
+        help="Manual: The accounting entries to value "
+             "the inventory are not posted automatically. \n"
```
Comment (elicoidal) [inline addons/stock_account/models/product.py]: s/posted/created

File: addons/stock_landed_costs/models/stock_landed_cost.py (hunk 1 above)
Comment (elicoidal) [inline addons/stock_landed_costs/models/stock_landed_cost.py]: @hbto 
> Landed cost is only to be applied on incoming moves,

Why? Technically a transport between warehouse would alter the value of the product. 
I know that the landing cost of the incoming shipment need to be accounted in the value of the product. I am not sure about other landed costs (warehouse transfers, relabels etc. happening afterwards)

**Comment by hbto** [comment]
Hi @elicoidal 


> Why? Technically a transport between warehouse would alter the value of the product.
> I know that the landing cost of the incoming shipment need to be accounted in the value of the product. I am not sure about other landed costs (warehouse transfers, relabels etc. happening afterwards)

In that case, Previously, prior to this PR, we have that move to/from  `transit` were considered outs/in, 

in the case you are pointing out the new approach will not work for reevaluating the product properly as revaluation is only counted on ins/outs, isn't it? @jco-odoo @sle-odoo 

Regards

**Comment by sle-odoo** [comment]
superseded by #19049
